### PR TITLE
Split public rooms directory auth config in two

### PR DESCRIPTION
**Pull request by babolivier**
Split the existing `restrict_public_rooms_to_local_users` config in two, effectively making `restrict_public_rooms_to_local_users` handle denying public rooms directory requests over federation only, and adding `restrict_public_rooms_to_registered_users` to require auth on client-side requests to `/publicRooms`.

Fixes https://github.com/matrix-org/synapse/issues/5532

### CODE REVIEW

**Comment by erikjohnston** [comment]
This changes the semantics of the current config options and will change existing deployments that have `restrict_public_rooms_to_local_users` set to true to start allowing client requests, which seems bad.

**Comment by babolivier** [comment]
Right, so the strategy we're going with right now is:

* use two new config flags (as in, not reusing an existing one), which are `allow_public_rooms_without_auth` and `allow_public_rooms_over_federation`. Both default to `true` in order to keep the default behaviour as it was.
* check for the (now obsolete) `restrict_public_rooms_to_local_users` config flag, if this one is set to `true` then set both new options to `false`.
* use this opportunity to change the `restrict_*` into `allow_*` to better match the style of the other config options.

File: synapse/config/server.py [hunk 1]
```diff
@@ -82,12 +82,22 @@ def read_config(self, config, **kwargs):
...
-        # homeserver to fetch it via federation.
-        self.restrict_public_rooms_to_local_users = config.get(
-            "restrict_public_rooms_to_local_users", False
-        )
+        # Check if the legacy "restrict_public_rooms_to_local_users" flag is set. This
+        # flag is now obsolete but we need to check it for backward-compatibility.
+        if config.get("restrict_public_rooms_to_local_users", False):
+            self.allow_public_rooms_without_auth = False
+            self.allow_public_rooms_over_federation = False
+        else:
```
Comment (erikjohnston) [inline synapse/config/server.py]: Can we raise a `ConfigError` if we have both the old flag and new flags set?

**Comment by babolivier** [comment]
> could you document, somewhere that a user might see it (like the upgrade notes or the sample config), what will happen if you have set `restrict_public_rooms_to_local_users`?

I added this to the changelog.

**Comment by erikjohnston** [comment]
Functionally this looks fine, I'll leave it up to @richvdh if changelog is good enough or it should also go in upgrade notes

**Pull request merged**